test/integration/watch_cache_test.go

```diff
@@ -0,0 +1,113 @@
+package integration
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	configapi "github.com/openshift/origin/pkg/cmd/server/api"
+	"github.com/openshift/origin/pkg/util/restoptions"
+	testutil "github.com/openshift/origin/test/util"
+	testserver "github.com/openshift/origin/test/util/server"
+	coreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+)
+
+func testWatchCacheWithConfig(t *testing.T, master *configapi.MasterConfig, expectedCacheSize, counterExampleCacheSize int) {
+	clusterAdminKubeConfig, err := testserver.StartConfiguredMasterAPI(master)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	client, err := testutil.GetClusterAdminKubeClient(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// create client with high burst value, otherwise we can only do 5 changes per second
+	config, err := testutil.GetClusterAdminClientConfig(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	config.Burst = expectedCacheSize
+	patchClient, err := coreclient.NewForConfig(config)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// modify labels of default namespace expectedCacheSize + 1 times
+	defaultNS, err := client.Core().Namespaces().Get("default", metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	startVersion, err := strconv.Atoi(defaultNS.ResourceVersion)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	for i := 0; i < expectedCacheSize+10; i++ {
+		defaultNS, err = patchClient.Namespaces().Patch("default", types.StrategicMergePatchType,
+			[]byte(fmt.Sprintf(`{"metadata":{"labels":{"test":"%d"}}}`, i)))
+		if err != nil {
+			t.Fatalf("unexpected patch error: %v", err)
+		}
+		t.Logf("version %v", defaultNS.ResourceVersion)
+	}
+
+	// try watch with very old resource version, not really expectedCacheSize versions back (there
+	// might be other namespace changes which push the default namespace versions out of the cache.
+	// Also note that the resource versions are global in etcd. So other resources will also lead
+	// to resource version jumps.
+	rv, err := strconv.Atoi(defaultNS.ResourceVersion)
+	if err != nil {
+		t.Fatalf("unexpected error converting the resource version: %v", err)
+	}
+	w, err := client.Core().Namespaces().Watch(metav1.ListOptions{ResourceVersion: strconv.Itoa(rv - (expectedCacheSize-counterExampleCacheSize)/2)})
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	defer w.Stop()
+
+	// try watch with an version that is too old
+	rv, err = strconv.Atoi(defaultNS.ResourceVersion)
+	if err != nil {
+		t.Fatalf("unexpected error converting the resource version: %v", err)
+	}
+	w, err = client.Core().Namespaces().Watch(metav1.ListOptions{ResourceVersion: strconv.Itoa(startVersion - 1)})
```

Comment:
Alright, here's a theory. Our cache hasn't caught up yet?  Try a get with `GetOptions` set to your latest resource version before starting the watch.